.github/workflows/pr.yml

```diff
@@ -88,42 +101,47 @@ jobs:
-      #       not needed.
-      ### - name: Maven Cache
-      ###   id: maven-cache
-      ###   uses: actions/cache@v2
-      ###   with:
-      ###     path: ~/.m2/repository
-      ###     key: ctia-m2-cache-${{ hashFiles('project.clj') }}
-      ###     restore-keys: |
-      ###       ctia-m2-cache-
-      ### # warm maven dependency cache if needed
-      ### - name: Setup Java
-      ###   if: steps.maven-cache.cache-hit != 'true'
-      ###   uses: actions/setup-java@v1
-      ###   with:
-      ###     java-version: ${{ env.JAVA_VERSION }}
-      ### - run: java -version
-      ### # install lein using cache to avoid extra download and startup time
-      ### - name: Cache Leiningen
-      ###   if: steps.maven-cache.cache-hit != 'true'
-      ###   uses: actions/cache@v2
-      ###   id: lein-cache
-      ###   with:
-      ###     path: ${{ env.BIN_PATH }}/lein
-      ###     key: ctia-lein-cache-${{ env.LEIN_VERSION }}
-      ### - name: Install Leiningen
-      ###   if: steps.lein-cache.cache-hit != 'true' && steps.maven-cache.cache-hit != 'true'
-      ###   run: |
-      ###     cd "${BIN_PATH}"
-      ###     curl -sLO "https://raw.githubusercontent.com/technomancy/leiningen/${LEIN_VERSION}/bin/lein"
-      ###     chmod a+x lein
-      ###     lein version
-      ### - name: Warm dependency cache
-      ###   if: steps.maven-cache.cache-hit != 'true'
-      ###   run: lein deps
+      # warm dependency cache on project.clj change (ie., when cache-hit is not true).
+      - name: Maven Cache
+        id: maven-cache
+        uses: actions/cache@v2
+        with:
+          path: ~/.m2/repository
+          key: ctia-m2-cache-${{ env.ACTIONS_CACHE_VERSION }}-${{ hashFiles('project.clj') }}
+          # these will be considered a cache-miss by `cache_hit`
+          restore-keys: |
+            ctia-m2-cache-${{ env.ACTIONS_CACHE_VERSION }}-
+      - name: Setup Java
+        if: steps.maven-cache.outputs.cache-hit != 'true'
+        uses: actions/setup-java@v1
+        with:
+          java-version: ${{ env.JAVA_VERSION }}
+      - run: java -version
+        if: steps.maven-cache.outputs.cache-hit != 'true'
+      - name: Leiningen Cache
+        if: steps.maven-cache.outputs.cache-hit != 'true'
+        uses: actions/cache@v2
+        id: lein-cache
+        with:
+          # note: these should coincide with ${BIN_PATH}/lein and ${LEIN_HOME}. unclear
+          # if variables can be used here
+          path: |
+            ~/bin/lein
+            ~/.lein
+          key: ctia-lein-cache-${{ env.ACTIONS_CACHE_VERSION }}-${{ env.LEIN_VERSION }}
+      - name: Install Leiningen
+        if: steps.maven-cache.outputs.cache-hit != 'true' && steps.lein-cache.outputs.cache-hit != 'true'
+        run: |
+          cd "${BIN_PATH}"
+          # poor man's travis_retry
+          for i in {1..${ACTIONS_RETRY_TIMES}}; do curl -sLO "https://raw.githubusercontent.com/technomancy/leiningen/${LEIN_VERSION}/bin/lein" && break; done
```

Comment:
Merely curious, why are these Lein commands flaky?

Reply:
I've never observed lein downloads as flaky, these are looped mainly because it's in the category of "downloading stuff".